viamonvifdiscovery/service.go

```diff
@@ -0,0 +1,201 @@
+		}
+	}
+	return nil, nil
+}
+
+type rtspDiscovery struct {
+	resource.Named
+	resource.AlwaysRebuild
+	Credentials []Creds
+	logger      logging.Logger
+}
+
+func newDiscovery(_ context.Context, _ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+	dis := &rtspDiscovery{
+		Named: conf.ResourceName().AsNamed(),
+		// place the empty credentials first, so we find the insecure cameras before other cameras.
+		Credentials: append([]Creds{emptyCred}, cfg.Credentials...),
+		logger:      logger,
+	}
+
+	dis.setCredNumbers()
+	return dis, nil
+}
+
+// since camera logins can have the same username,
+// check how many repeated usernames we have and index them, to ensure cameras have unique names.
+func (dis *rtspDiscovery) setCredNumbers() {
+	// usernames do not have to be unique, so we want to add an additional label to ensure cameras have unique names.
+	for index, cred := range dis.Credentials {
+		for counter, otherCreds := range dis.Credentials {
+			if counter <= index {
+				continue
+			}
+			// increase the credNumber for later cameras.
+			if cred.Username == otherCreds.Username {
+				dis.Credentials[counter].credNumber++
+			}
+		}
+	}
+}
+
+// Close closes the discovery service.
+func (dis *rtspDiscovery) Close(_ context.Context) error {
+	return nil
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *rtspDiscovery) DiscoverResources(_ context.Context, _ map[string]any) ([]resource.Config, error) {
+	potentialCams := []resource.Config{}
+	insecureURLs := []string{}
+
+	for _, cred := range dis.Credentials {
+		// if we have multiple empty/insecure credentials configured for some reason, skip to the next credential.
+		// only skip if we have not processed insecure credentials yet.
+		if cred.isInsecure() && len(insecureURLs) > 0 {
+			continue
+		}
+
+		list, err := viamonvif.DiscoverCameras(cred.Username, cred.Password, dis.logger, nil)
+		if err != nil {
+			return nil, err
+		}
+		cameraNumber := 0
+		for _, l := range list.Cameras {
+			dis.logger.Debugf("%s %s %s", l.Manufacturer, l.Model, l.SerialNumber)
+			// some cameras return with no urls. explicitly skipping those so the behavior is clear in the service.
+			if len(l.RTSPURLs) == 0 {
+				dis.logger.Debugf("No urls found for camera, skipping. %s %s %s", l.Manufacturer, l.Model, l.SerialNumber)
+				continue
+			}
+			cameraNumber++
+			camConfigs, err := createCamerasFromURLs(cred.createName(cameraNumber), l, insecureURLs, dis.logger)
+			if err != nil {
+				return nil, err
+			}
+			potentialCams = append(potentialCams, camConfigs...)
+
+			// if we are processing insecure creds, record the urls so we do not duplicate the cameras when processing cameras with creds.
+			if cred.isInsecure() {
+				insecureURLs = append(insecureURLs, l.NoLoginURLs...)
+			}
```

Comment:
I'm sort of cofnused about the flow of insecureCreds - we check that the rtspStream requires no credentials, then de-dupe the camera names, but were does this list of insecure URLs go?

Reply:
At the time we would compare the current url with the list of insecure urls, and skip creating the camera if it was found in the list.

this logic has been removed